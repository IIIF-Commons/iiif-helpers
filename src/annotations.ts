import { ChoiceBody, ImageService, InternationalString } from '@iiif/presentation-3';
import { ParsedSelector, SupportedTarget } from './annotation-targets';
import { ChoiceDescription, ComplexChoice, SingleChoice } from './painting-annotations';

// A wrapper for annotations that are "well-known"
interface ClassifiedAnnotation {
  id: string;
  type: 'Annotation';
  label?: InternationalString;
  summary?: InternationalString;
  motivation: string;
  target: SupportedTarget;
  body: SupportedBody;
}

interface AnnotationClassificationOptions {
  language?: string;
  choices?: string[];
  isTrusted?: boolean;
}

interface SimpleImage {
  id: string;
  format: string;
  label?: InternationalString;
  summary?: InternationalString;
  width?: number;
  height?: number;
  service?: ImageService;
}

type SupportedBody = {
  type: string;
  order: Array<'label' | 'summary' | 'image' | 'link' | 'internalLink' | 'content'>; // e.g. ['label', 'image', 'link', 'image']
  label?: InternationalString;
  summary?: InternationalString;
  content: InternationalString;
  image?: SimpleImage;
  images?: SimpleImage[];
  internalLink: {
    id: string;
    label?: InternationalString;
    type: 'Canvas';
    target: ParsedSelector;
  } | null;
  link: {
    id: string;
    label?: InternationalString;
    language?: string;
  } | null;
  contentList: InternationalString[];
  rawBody: any[];
  containsHtml: boolean;
  choice: ChoiceDescription | null;
};
