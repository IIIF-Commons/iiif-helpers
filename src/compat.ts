export type CompatVault = {
  get: import('@iiif/vault').Vault['get'];
};

export const compatVault: CompatVault = {
  get(nonRef: any) {
    return nonRef;
  },
};
