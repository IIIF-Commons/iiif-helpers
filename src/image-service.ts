export * from './image-service/image-service-loader';
export * from './image-service/get-image-server-from-id';
export * from './image-service/sampled-tiles-to-tiles';
export * from './image-service/pick-best-from-candidates';
export * from './image-service/is-best-match';
export * from './image-service/image-service-store';
export * from './image-service/get-image-from-tile-source';
export * from './image-service/get-image-candidates';
export * from './image-service/get-image-candidates-from-service';
export * from './image-service/get-fixed-sizes-from-service';
export * from './image-service/is-image-3';
export * from './image-service/get-custom-size-from-service';
export * from './image-service/get-fixed-size-from-image';
export * from './image-service/infer-size-from-url';
export * from './image-service/image-sizes-match';
export * from './image-service/get-smallest-scale-factor-as-single-image';
export * from './image-service/types';
export * from './image-service/image-service-try';
