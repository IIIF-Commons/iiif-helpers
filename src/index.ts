export * from './events';
export * from './styles';
export * from './thumbnail';
export * from './i18n';
