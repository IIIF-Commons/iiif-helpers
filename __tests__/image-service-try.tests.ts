import mitt from "mitt";
import { ImageServer, createImageServiceStore, createImageServiceTryStore } from "../src/image-service";
import { describe, expect, test } from 'vitest';
import { ImageService } from "@iiif/presentation-3";
import { generateServiceFromImageUrl } from "../src/image-service/generate-service-from-image-url";

describe("ImageServiceTry", () => {

  test('Example image service try', () => {
    const dlcs = {
      "@context": "http://iiif.io/api/image/2/context.json",
      "@id": "https://dlc.services/thumbs/7/18/9a4356aa-2953-4ce8-9b94-ec8b10bf7921",
      "@type": "iiif:Image",
      "profile": [
        "http://iiif.io/api/image/2/level0.json",
        {
          "formats": [
            "jpg"
          ],
          "qualities": [
            "color"
          ],
          "supports": [
            "sizeByWhListed"
          ]
        }
      ],
      "protocol": "http://iiif.io/api/image",
      "width": 1024,
      "height": 683,
      "sizes": [
        {
          "width": 100,
          "height": 67
        },
        {
          "width": 200,
          "height": 133
        },
        {
          "width": 400,
          "height": 267
        },
        {
          "width": 1024,
          "height": 683
        }
      ]
    };

    const loader = createImageServiceStore();
    const store = createImageServiceTryStore({
      loader: loader.store,
      events: mitt(),
      imageService: dlcs as ImageService
    });

    store.getState().try({ minWidth: 100000, height: 267 });
    store.getState().try({ width: 400, height: 267 });
    store.getState().try({ width: 200, height: 133 });
    store.getState().try({ width: 100, height: 67 });

    expect(
      store.getState().tryCursor
    ).toMatchInlineSnapshot(`1`);

    expect(
      store.getState().image()
    ).toMatchInlineSnapshot(`"https://dlc.services/thumbs/7/18/9a4356aa-2953-4ce8-9b94-ec8b10bf7921/full/400,/0/default.jpg"`);

    const { sizeByWhListed } = store.getState();

    const sizes = store.getState().sizes;
    expect(sizes).toEqual([
      { width: 100, height: 67 },
      { width: 200, height: 133 },
      { width: 400, height: 267 },
      { width: 1024, height: 683 }
    ]);

    expect(
      sizeByWhListed(sizes[0])
    ).toMatchInlineSnapshot(`"https://dlc.services/thumbs/7/18/9a4356aa-2953-4ce8-9b94-ec8b10bf7921/full/100,/0/default.jpg"`);
    expect(
      sizeByWhListed(sizes[1])
    ).toMatchInlineSnapshot(`"https://dlc.services/thumbs/7/18/9a4356aa-2953-4ce8-9b94-ec8b10bf7921/full/200,/0/default.jpg"`);
    expect(
      sizeByWhListed(sizes[2])
    ).toMatchInlineSnapshot(`"https://dlc.services/thumbs/7/18/9a4356aa-2953-4ce8-9b94-ec8b10bf7921/full/400,/0/default.jpg"`);
    expect(
      sizeByWhListed(sizes[3])
    ).toMatchInlineSnapshot(`"https://dlc.services/thumbs/7/18/9a4356aa-2953-4ce8-9b94-ec8b10bf7921/full/1024,/0/default.jpg"`);

    expect(store.getState()).toMatchSnapshot();
  });

  test('nls example', () => {
    const service = {
      "@context": "http://iiif.io/api/image/2/context.json",
      "@id": "https://view.nls.uk/iiif/7443/74438559.5",
      "profile": "http://iiif.io/api/image/2/profiles/level2.json",
      "height": 1781,
      "width": 2500
    };

    const loader = createImageServiceStore();
    const store = createImageServiceTryStore({
      loader: loader.store,
      events: mitt(),
      imageService: service as ImageService
    });


    const { generateImageUrlAtSize, ingestImageService, formats, level } = store.getState();

    expect(level).toBe(2);
    expect(formats).toMatchInlineSnapshot(`
      [
        "jpg",
        "png",
      ]
    `);

    expect(
      generateImageUrlAtSize({ width: 250 })
    ).toBe("https://view.nls.uk/iiif/7443/74438559.5/full/250,/0/default.jpg");

    // Now the full service.
    ingestImageService({
      "@context": "http://iiif.io/api/image/2/context.json",
      "@id": "https://view.nls.uk/iiif/7443/74438559.5",
      "protocol": "http://iiif.io/api/image",
      "width": 2500,
      "height": 1778,
      "tiles": [
        {
          "width": 256,
          "height": 256,
          "scaleFactors": [
            1,
            2,
            4,
            8,
            16
          ]
        }
      ],
      "profile": [
        "http://iiif.io/api/image/2/level1.json",
        {
          "formats": [
            "jpg"
          ],
          "qualities": [
            "native",
            "color",
            "gray"
          ],
          "supports": [
            "regionByPct",
            "sizeByForcedWh",
            "sizeByWh",
            "sizeAboveFull",
            "rotationBy90s",
            "mirroring",
            "gray"
          ]
        }
      ]
    } as ImageService);

    expect(store.getState()).toMatchSnapshot();
  });


  test('Fixture - level 0, no thumbnail', () => {

    const service1 = {
      "@context": "http://iiif.io/api/image/2/context.json",
      "@id": "https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/no_thum_prop/v2/level0/tractor",
      "profile": "http://iiif.io/api/image/2/level0.json"
    };
    const service1Full = {
      "tiles": [
        {
          "scaleFactors": [
            32,
            16,
            8,
            4,
            2,
            1
          ],
          "width": 1024,
          "height": 1024
        }
      ],
      "protocol": "http://iiif.io/api/image",
      "sizes": [
        {
          "width": 126,
          "height": 95
        },
        {
          "width": 252,
          "height": 189
        },
        {
          "width": 504,
          "height": 378
        },
        {
          "width": 1008,
          "height": 756
        },
        {
          "width": 2016,
          "height": 1512
        },
        {
          "width": 4031,
          "height": 3023
        }
      ],
      "profile": "http://iiif.io/api/image/2/level0.json",
      "width": 4031,
      "@id": "https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/no_thum_prop/v2/level0/tractor",
      "@context": "http://iiif.io/api/image/2/context.json",
      "height": 3023
    };


    const loader = createImageServiceStore();
    const store = createImageServiceTryStore({
      loader: loader.store,
      events: mitt(),
      imageService: service1 as ImageService
    });

    const { fullTiles, ingestImageService, sizeByWhListed, features } = store.getState();

    expect(fullTiles).toMatchInlineSnapshot(`[]`);
    expect(features).toMatchInlineSnapshot(`
      {
        "cors": false,
        "corsVerified": false,
        "sizeByH": false,
        "sizeByW": false,
        "sizeByWh": false,
        "sizeByWhListed": true,
      }
    `);

    ingestImageService(service1Full as ImageService);

    expect(store.getState()).toMatchSnapshot();

    expect(
      sizeByWhListed({ width: 126, height: 95 })
    ).toMatchInlineSnapshot(`"https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/no_thum_prop/v2/level0/tractor/full/126,/0/default.jpg"`);

    expect(
      sizeByWhListed({ width: 252, height: 189 })
    ).toMatchInlineSnapshot(`"https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/no_thum_prop/v2/level0/tractor/full/252,/0/default.jpg"`);

    expect(
      sizeByWhListed({ width: 504, height: 378 })
    ).toMatchInlineSnapshot(`"https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/no_thum_prop/v2/level0/tractor/full/504,/0/default.jpg"`);

    expect(
      sizeByWhListed({ width: 1008, height: 756 })
    ).toMatchInlineSnapshot(`"https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/no_thum_prop/v2/level0/tractor/full/1008,/0/default.jpg"`);
  });

  test('Fixture - level 0 thumbnail', () => {
    const service = {
      "protocol": "http://iiif.io/api/image",
      "sizes": [
        {
          "width": 170,
          "height": 226
        },
        {
          "width": 340,
          "height": 452
        },
        {
          "width": 679,
          "height": 904
        },
        {
          "width": 1357,
          "height": 1808
        },
        {
          "width": 2714,
          "height": 3615
        },
        {
          "width": 5428,
          "height": 7230
        }
      ],
      "profile": "http://iiif.io/api/image/2/level0.json",
      "width": 5428,
      "@id": "https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/thum_prop/v2/canvas_thum_level0/titlepage1_verso",
      "@context": "http://iiif.io/api/image/2/context.json",
      "height": 7230
    };


    const loader = createImageServiceStore();
    const store = createImageServiceTryStore({
      loader: loader.store,
      events: mitt(),
      imageService: service as ImageService
    });

    expect(
      store.getState()
    ).toMatchSnapshot();

    const { sizeByWhListed } = store.getState();

    expect(
      sizeByWhListed({ width: 170, height: 226 })
    ).toMatchInlineSnapshot(`"https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/thum_prop/v2/canvas_thum_level0/titlepage1_verso/full/170,/0/default.jpg"`);

    expect(
      sizeByWhListed({ width: 340, height: 452 })
    ).toMatchInlineSnapshot(`"https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/thum_prop/v2/canvas_thum_level0/titlepage1_verso/full/340,/0/default.jpg"`);

    expect(
      sizeByWhListed({ width: 679, height: 904 })
    ).toMatchInlineSnapshot(`"https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/thum_prop/v2/canvas_thum_level0/titlepage1_verso/full/679,/0/default.jpg"`);
  });


  test('Situation with thumbnail where we only have a string', () => {

    const thumbnailString = 'https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-4_titlepage1_verso/full/512,/0/default.jpg';

    const service = generateServiceFromImageUrl(thumbnailString, {
      "height": 7230,
      "width": 5428,
    });

    expect(service).toMatchInlineSnapshot(`
      {
        "@context": "http://iiif.io/api/image/2/context.json",
        "@id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-4_titlepage1_verso",
        "@type": "ImageService2",
        "profile": "http://iiif.io/api/image/2/level0.json",
        "protocol": "http://iiif.io/api/image",
        "sizes": [
          {
            "height": 682,
            "width": 512,
          },
        ],
      }
    `);

    const loader = createImageServiceStore();
    const store = createImageServiceTryStore({
      loader: loader.store,
      events: mitt(),
      imageService: service as ImageService
    });

    expect(store.getState()).toMatchSnapshot();

    const { sizeByWhListed } = store.getState();

    expect(
      sizeByWhListed({ width: 512, height: 682 })
    ).toEqual(thumbnailString);

    const fullService: ImageService = {
      "@context": "http://iiif.io/api/image/2/context.json",
      "@id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-4_titlepage1_verso",
      "height": 7230,
      "profile": [
        "http://iiif.io/api/image/2/level1.json",
        {
          "formats": [
            "jpg",
            "png"
          ],
          "qualities": [
            "default",
            "color",
            "gray"
          ]
        }
      ],
      "protocol": "http://iiif.io/api/image",
      "tiles": [
        {
          "height": 512,
          "scaleFactors": [
            1,
            2,
            4,
            8
          ],
          "width": 512
        }
      ],
      "width": 5428
    };

    store.getState().ingestImageService(fullService);

    expect(store.getState()).toMatchSnapshot();

  })
});
