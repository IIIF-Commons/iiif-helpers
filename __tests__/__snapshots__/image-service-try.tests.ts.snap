// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ImageServiceTry > Example image service try 7`] = `
{
  "asyncImage": [Function],
  "asyncTry": [Function],
  "constrain": [Function],
  "constraints": {
    "format": "jpg",
    "height": 256,
    "maxArea": Infinity,
    "maxHeight": Infinity,
    "maxWidth": Infinity,
    "minArea": 0,
    "minHeight": 0,
    "minWidth": 0,
    "profile": null,
    "quality": "default",
    "region": {
      "full": true,
    },
    "rotation": {
      "angle": 0,
    },
    "width": 256,
  },
  "defaultImage": null,
  "didError": false,
  "error": null,
  "fallback": null,
  "features": {
    "cors": false,
    "corsVerified": false,
    "sizeByH": false,
    "sizeByW": false,
    "sizeByWh": false,
    "sizeByWhListed": true,
  },
  "formats": [
    "jpg",
  ],
  "fullService": null,
  "fullTiles": [],
  "generateImageUrlAtSize": [Function],
  "image": [Function],
  "imageServer": {
    "path": "thumbs/7/18/9a4356aa-2953-4ce8-9b94-ec8b10bf7921/info.json",
    "prefix": "",
    "scheme": "https",
    "server": "dlc.services",
  },
  "imageServiceInfoJson": "https://dlc.services/thumbs/7/18/9a4356aa-2953-4ce8-9b94-ec8b10bf7921/info.json",
  "ingestImageService": [Function],
  "isFullyLoaded": false,
  "isLoading": false,
  "level": 0,
  "maxArea": Infinity,
  "maxHeight": 683,
  "maxWidth": 1024,
  "options": {
    "cors": false,
    "headers": {},
    "placeholder": null,
    "rounding": "ceil",
    "sizePriority": "largest-size",
    "stitch": false,
    "unsafe": true,
  },
  "originalService": {
    "@context": "http://iiif.io/api/image/2/context.json",
    "@id": "https://dlc.services/thumbs/7/18/9a4356aa-2953-4ce8-9b94-ec8b10bf7921",
    "@type": "iiif:Image",
    "height": 683,
    "profile": [
      "http://iiif.io/api/image/2/level0.json",
      {
        "formats": [
          "jpg",
        ],
        "qualities": [
          "color",
        ],
        "supports": [
          "sizeByWhListed",
        ],
      },
    ],
    "protocol": "http://iiif.io/api/image",
    "sizes": [
      {
        "height": 67,
        "width": 100,
      },
      {
        "height": 133,
        "width": 200,
      },
      {
        "height": 267,
        "width": 400,
      },
      {
        "height": 683,
        "width": 1024,
      },
    ],
    "width": 1024,
  },
  "qualities": [
    "default",
  ],
  "requestImageService": [Function],
  "sizeByH": [Function],
  "sizeByW": [Function],
  "sizeByWh": [Function],
  "sizeByWhListed": [Function],
  "sizes": [
    {
      "height": 67,
      "width": 100,
    },
    {
      "height": 133,
      "width": 200,
    },
    {
      "height": 267,
      "width": 400,
    },
    {
      "height": 683,
      "width": 1024,
    },
  ],
  "supportsArbitrarySizes": false,
  "supportsSizeByWhListed": false,
  "test": [Function],
  "tileSizes": [],
  "tiles": undefined,
  "try": [Function],
  "tryCursor": 1,
  "tryQueue": [
    {
      "constraints": {
        "height": 267,
        "minWidth": 100000,
      },
      "error": "Unknown error (2)",
      "id": null,
      "isAsync": false,
      "isError": true,
      "isFetching": false,
      "isVerified": false,
      "isVerifying": false,
    },
    {
      "constraints": {
        "height": 267,
        "width": 400,
      },
      "error": null,
      "id": "https://dlc.services/thumbs/7/18/9a4356aa-2953-4ce8-9b94-ec8b10bf7921/full/400,/0/default.jpg",
      "isAsync": false,
      "isError": false,
      "isFetching": false,
      "isVerified": false,
      "isVerifying": false,
    },
    {
      "constraints": {
        "height": 133,
        "width": 200,
      },
      "error": null,
      "id": "https://dlc.services/thumbs/7/18/9a4356aa-2953-4ce8-9b94-ec8b10bf7921/full/200,/0/default.jpg",
      "isAsync": false,
      "isError": false,
      "isFetching": false,
      "isVerified": false,
      "isVerifying": false,
    },
    {
      "constraints": {
        "height": 67,
        "width": 100,
      },
      "error": null,
      "id": "https://dlc.services/thumbs/7/18/9a4356aa-2953-4ce8-9b94-ec8b10bf7921/full/100,/0/default.jpg",
      "isAsync": false,
      "isError": false,
      "isFetching": false,
      "isVerified": false,
      "isVerifying": false,
    },
  ],
  "version": 2,
}
`;

exports[`ImageServiceTry > Fixture - level 0 thumbnail 1`] = `
{
  "asyncImage": [Function],
  "asyncTry": [Function],
  "constrain": [Function],
  "constraints": {
    "format": "jpg",
    "height": 256,
    "maxArea": Infinity,
    "maxHeight": Infinity,
    "maxWidth": Infinity,
    "minArea": 0,
    "minHeight": 0,
    "minWidth": 0,
    "profile": null,
    "quality": "default",
    "region": {
      "full": true,
    },
    "rotation": {
      "angle": 0,
    },
    "width": 256,
  },
  "defaultImage": null,
  "didError": false,
  "error": null,
  "fallback": null,
  "features": {
    "cors": false,
    "corsVerified": false,
    "sizeByH": false,
    "sizeByW": false,
    "sizeByWh": false,
    "sizeByWhListed": true,
  },
  "formats": [
    "jpg",
  ],
  "fullService": null,
  "fullTiles": [],
  "generateImageUrlAtSize": [Function],
  "image": [Function],
  "imageServer": {
    "path": "fixtures/examples/thumbnail/canvas/thum_prop/v2/canvas_thum_level0/titlepage1_verso/info.json",
    "prefix": "",
    "scheme": "https",
    "server": "iiif-commons.github.io",
  },
  "imageServiceInfoJson": "https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/thum_prop/v2/canvas_thum_level0/titlepage1_verso/info.json",
  "ingestImageService": [Function],
  "isFullyLoaded": false,
  "isLoading": false,
  "level": 0,
  "maxArea": Infinity,
  "maxHeight": 7230,
  "maxWidth": 5428,
  "options": {
    "cors": false,
    "headers": {},
    "placeholder": null,
    "rounding": "ceil",
    "sizePriority": "largest-size",
    "stitch": false,
    "unsafe": true,
  },
  "originalService": {
    "@context": "http://iiif.io/api/image/2/context.json",
    "@id": "https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/thum_prop/v2/canvas_thum_level0/titlepage1_verso",
    "height": 7230,
    "profile": "http://iiif.io/api/image/2/level0.json",
    "protocol": "http://iiif.io/api/image",
    "sizes": [
      {
        "height": 226,
        "width": 170,
      },
      {
        "height": 452,
        "width": 340,
      },
      {
        "height": 904,
        "width": 679,
      },
      {
        "height": 1808,
        "width": 1357,
      },
      {
        "height": 3615,
        "width": 2714,
      },
      {
        "height": 7230,
        "width": 5428,
      },
    ],
    "width": 5428,
  },
  "qualities": [
    "default",
  ],
  "requestImageService": [Function],
  "sizeByH": [Function],
  "sizeByW": [Function],
  "sizeByWh": [Function],
  "sizeByWhListed": [Function],
  "sizes": [
    {
      "height": 226,
      "width": 170,
    },
    {
      "height": 452,
      "width": 340,
    },
    {
      "height": 904,
      "width": 679,
    },
    {
      "height": 1808,
      "width": 1357,
    },
    {
      "height": 3615,
      "width": 2714,
    },
    {
      "height": 7230,
      "width": 5428,
    },
  ],
  "supportsArbitrarySizes": false,
  "supportsSizeByWhListed": false,
  "test": [Function],
  "tileSizes": [],
  "tiles": undefined,
  "try": [Function],
  "tryCursor": -1,
  "tryQueue": [],
  "version": 2,
}
`;

exports[`ImageServiceTry > Fixture - level 0, no thumbnail 3`] = `
{
  "asyncImage": [Function],
  "asyncTry": [Function],
  "constrain": [Function],
  "constraints": {
    "format": "jpg",
    "height": 256,
    "maxArea": Infinity,
    "maxHeight": Infinity,
    "maxWidth": Infinity,
    "minArea": 0,
    "minHeight": 0,
    "minWidth": 0,
    "profile": null,
    "quality": "default",
    "region": {
      "full": true,
    },
    "rotation": {
      "angle": 0,
    },
    "width": 256,
  },
  "defaultImage": null,
  "didError": false,
  "error": null,
  "fallback": null,
  "features": {
    "cors": false,
    "corsVerified": false,
    "sizeByH": false,
    "sizeByW": false,
    "sizeByWh": false,
    "sizeByWhListed": true,
  },
  "formats": [
    "jpg",
  ],
  "fullService": null,
  "fullTiles": [
    {
      "height": 95,
      "scale": 32,
      "width": 126,
    },
    {
      "height": 189,
      "scale": 16,
      "width": 252,
    },
    {
      "height": 378,
      "scale": 8,
      "width": 504,
    },
    {
      "height": 756,
      "scale": 4,
      "width": 1008,
    },
  ],
  "generateImageUrlAtSize": [Function],
  "image": [Function],
  "imageServer": {
    "path": "fixtures/examples/thumbnail/canvas/no_thum_prop/v2/level0/tractor/info.json",
    "prefix": "",
    "scheme": "https",
    "server": "iiif-commons.github.io",
  },
  "imageServiceInfoJson": "https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/no_thum_prop/v2/level0/tractor/info.json",
  "ingestImageService": [Function],
  "isFullyLoaded": false,
  "isLoading": false,
  "level": 0,
  "maxArea": Infinity,
  "maxHeight": 3023,
  "maxWidth": 4031,
  "options": {
    "cors": false,
    "headers": {},
    "placeholder": null,
    "rounding": "ceil",
    "sizePriority": "largest-size",
    "stitch": false,
    "unsafe": true,
  },
  "originalService": {
    "@context": "http://iiif.io/api/image/2/context.json",
    "@id": "https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/no_thum_prop/v2/level0/tractor",
    "profile": "http://iiif.io/api/image/2/level0.json",
  },
  "qualities": [
    "default",
  ],
  "requestImageService": [Function],
  "sizeByH": [Function],
  "sizeByW": [Function],
  "sizeByWh": [Function],
  "sizeByWhListed": [Function],
  "sizes": [
    {
      "height": 95,
      "width": 126,
    },
    {
      "height": 189,
      "width": 252,
    },
    {
      "height": 378,
      "width": 504,
    },
    {
      "height": 756,
      "width": 1008,
    },
    {
      "height": 1512,
      "width": 2016,
    },
    {
      "height": 3023,
      "width": 4031,
    },
  ],
  "supportsArbitrarySizes": false,
  "supportsSizeByWhListed": false,
  "test": [Function],
  "tileSizes": [
    1024,
  ],
  "tiles": [
    {
      "height": 1024,
      "scaleFactors": [
        32,
        16,
        8,
        4,
        2,
        1,
      ],
      "width": 1024,
    },
  ],
  "try": [Function],
  "tryCursor": -1,
  "tryQueue": [],
  "version": 2,
}
`;

exports[`ImageServiceTry > Situation with thumbnail where we only have a string 2`] = `
{
  "asyncImage": [Function],
  "asyncTry": [Function],
  "constrain": [Function],
  "constraints": {
    "format": "jpg",
    "height": 256,
    "maxArea": Infinity,
    "maxHeight": Infinity,
    "maxWidth": Infinity,
    "minArea": 0,
    "minHeight": 0,
    "minWidth": 0,
    "profile": null,
    "quality": "default",
    "region": {
      "full": true,
    },
    "rotation": {
      "angle": 0,
    },
    "width": 256,
  },
  "defaultImage": null,
  "didError": false,
  "error": null,
  "fallback": null,
  "features": {
    "cors": false,
    "corsVerified": false,
    "sizeByH": false,
    "sizeByW": false,
    "sizeByWh": false,
    "sizeByWhListed": true,
  },
  "formats": [
    "jpg",
  ],
  "fullService": null,
  "fullTiles": [],
  "generateImageUrlAtSize": [Function],
  "image": [Function],
  "imageServer": {
    "path": "api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-4_titlepage1_verso/info.json",
    "prefix": "",
    "scheme": "https",
    "server": "iiif.io",
  },
  "imageServiceInfoJson": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-4_titlepage1_verso/info.json",
  "ingestImageService": [Function],
  "isFullyLoaded": false,
  "isLoading": false,
  "level": 0,
  "maxArea": Infinity,
  "maxHeight": Infinity,
  "maxWidth": Infinity,
  "options": {
    "cors": false,
    "headers": {},
    "placeholder": null,
    "rounding": "ceil",
    "sizePriority": "largest-size",
    "stitch": false,
    "unsafe": true,
  },
  "originalService": {
    "@context": "http://iiif.io/api/image/2/context.json",
    "@id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-4_titlepage1_verso",
    "@type": "ImageService2",
    "profile": "http://iiif.io/api/image/2/level0.json",
    "protocol": "http://iiif.io/api/image",
    "sizes": [
      {
        "height": 682,
        "width": 512,
      },
    ],
  },
  "qualities": [
    "default",
  ],
  "requestImageService": [Function],
  "sizeByH": [Function],
  "sizeByW": [Function],
  "sizeByWh": [Function],
  "sizeByWhListed": [Function],
  "sizes": [
    {
      "height": 682,
      "width": 512,
    },
  ],
  "supportsArbitrarySizes": false,
  "supportsSizeByWhListed": false,
  "test": [Function],
  "tileSizes": [],
  "tiles": undefined,
  "try": [Function],
  "tryCursor": -1,
  "tryQueue": [],
  "version": 2,
}
`;

exports[`ImageServiceTry > Situation with thumbnail where we only have a string 3`] = `
{
  "asyncImage": [Function],
  "asyncTry": [Function],
  "constrain": [Function],
  "constraints": {
    "format": "jpg",
    "height": 256,
    "maxArea": Infinity,
    "maxHeight": Infinity,
    "maxWidth": Infinity,
    "minArea": 0,
    "minHeight": 0,
    "minWidth": 0,
    "profile": null,
    "quality": "default",
    "region": {
      "full": true,
    },
    "rotation": {
      "angle": 0,
    },
    "width": 256,
  },
  "defaultImage": null,
  "didError": false,
  "error": null,
  "fallback": null,
  "features": {
    "cors": true,
    "corsVerified": false,
    "sizeByH": true,
    "sizeByW": true,
    "sizeByWh": true,
    "sizeByWhListed": true,
  },
  "formats": [
    "jpg",
  ],
  "fullService": null,
  "fullTiles": [],
  "generateImageUrlAtSize": [Function],
  "image": [Function],
  "imageServer": {
    "path": "api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-4_titlepage1_verso/info.json",
    "prefix": "",
    "scheme": "https",
    "server": "iiif.io",
  },
  "imageServiceInfoJson": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-4_titlepage1_verso/info.json",
  "ingestImageService": [Function],
  "isFullyLoaded": false,
  "isLoading": false,
  "level": 1,
  "maxArea": Infinity,
  "maxHeight": 7230,
  "maxWidth": 5428,
  "options": {
    "cors": false,
    "headers": {},
    "placeholder": null,
    "rounding": "ceil",
    "sizePriority": "largest-size",
    "stitch": false,
    "unsafe": true,
  },
  "originalService": {
    "@context": "http://iiif.io/api/image/2/context.json",
    "@id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-4_titlepage1_verso",
    "@type": "ImageService2",
    "profile": "http://iiif.io/api/image/2/level0.json",
    "protocol": "http://iiif.io/api/image",
    "sizes": [
      {
        "height": 682,
        "width": 512,
      },
    ],
  },
  "qualities": [
    "default",
  ],
  "requestImageService": [Function],
  "sizeByH": [Function],
  "sizeByW": [Function],
  "sizeByWh": [Function],
  "sizeByWhListed": [Function],
  "sizes": undefined,
  "supportsArbitrarySizes": false,
  "supportsSizeByWhListed": false,
  "test": [Function],
  "tileSizes": [
    512,
  ],
  "tiles": [
    {
      "height": 512,
      "scaleFactors": [
        1,
        2,
        4,
        8,
      ],
      "width": 512,
    },
  ],
  "try": [Function],
  "tryCursor": -1,
  "tryQueue": [],
  "version": 2,
}
`;

exports[`ImageServiceTry > nls example 2`] = `
{
  "asyncImage": [Function],
  "asyncTry": [Function],
  "constrain": [Function],
  "constraints": {
    "format": "jpg",
    "height": 256,
    "maxArea": Infinity,
    "maxHeight": Infinity,
    "maxWidth": Infinity,
    "minArea": 0,
    "minHeight": 0,
    "minWidth": 0,
    "profile": null,
    "quality": "default",
    "region": {
      "full": true,
    },
    "rotation": {
      "angle": 0,
    },
    "width": 256,
  },
  "defaultImage": null,
  "didError": false,
  "error": null,
  "fallback": null,
  "features": {
    "cors": true,
    "corsVerified": false,
    "sizeByH": true,
    "sizeByW": true,
    "sizeByWh": true,
    "sizeByWhListed": true,
  },
  "formats": [
    "jpg",
  ],
  "fullService": null,
  "fullTiles": [
    {
      "height": 112,
      "scale": 16,
      "width": 157,
    },
  ],
  "generateImageUrlAtSize": [Function],
  "image": [Function],
  "imageServer": {
    "path": "iiif/7443/74438559.5/info.json",
    "prefix": "",
    "scheme": "https",
    "server": "view.nls.uk",
  },
  "imageServiceInfoJson": "https://view.nls.uk/iiif/7443/74438559.5/info.json",
  "ingestImageService": [Function],
  "isFullyLoaded": false,
  "isLoading": false,
  "level": 1,
  "maxArea": Infinity,
  "maxHeight": 1778,
  "maxWidth": 2500,
  "options": {
    "cors": false,
    "headers": {},
    "placeholder": null,
    "rounding": "ceil",
    "sizePriority": "largest-size",
    "stitch": false,
    "unsafe": true,
  },
  "originalService": {
    "@context": "http://iiif.io/api/image/2/context.json",
    "@id": "https://view.nls.uk/iiif/7443/74438559.5",
    "height": 1781,
    "profile": "http://iiif.io/api/image/2/profiles/level2.json",
    "width": 2500,
  },
  "qualities": [
    "default",
  ],
  "requestImageService": [Function],
  "sizeByH": [Function],
  "sizeByW": [Function],
  "sizeByWh": [Function],
  "sizeByWhListed": [Function],
  "sizes": undefined,
  "supportsArbitrarySizes": false,
  "supportsSizeByWhListed": false,
  "test": [Function],
  "tileSizes": [
    256,
  ],
  "tiles": [
    {
      "height": 256,
      "scaleFactors": [
        1,
        2,
        4,
        8,
        16,
      ],
      "width": 256,
    },
  ],
  "try": [Function],
  "tryCursor": -1,
  "tryQueue": [],
  "version": 2,
}
`;
